[FIRRTL] Verify that NodeOp is passive

seldridge opened the pull request:
Add a NodeOp verifier that asserts that a node is passive. One
testcase is added and one testcase is changed that did not respect
this invariant.

- Fixes #395.
- Fixes #388.

lattner commented:
nice, thank you for fixing this!  Does the firparser need to be updated to introduce an "aspassive" node for inputs that happen to be flip types?

Also, is it possible to handle this with a type constraint instead of a verifier hook?  Plz try:

`let arguments = (ins PassiveType:$input, OptionalAttr<StrAttr>:$name);`

in the NodeOp definition in the .td file, and see if you can drop the verifier hook.  Thanks!

darthscsi commented:
LGTM

seldridge commented:
@lattner wrote:
> Does the firparser need to be updated to introduce an "aspassive" node for inputs that happen to be flip types?

Yes, it does. I fixed this in #381, but I pulled out that portion into this PR in the rebase.

@lattner wrote:
> Also, is it possible to handle this with a type constraint instead of a verifier hook?

It is! Great idea. Updated to use this and I dropped the verifier hook.

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -2018,6 +2019,25 @@ ParseResult FIRStmtParser::parseNode() {
+  auto flip = initializerType.dyn_cast_or_null<FlipType>();
+  if (flip)
+    underlyingType = flip.getElementType();
+  if (!underlyingType.isPassive()) {
```
Because of the canonicalization rules for types (whereby the body of a flip is always passive), I think this is equivalent to:

`if (!initializerType.isPassiveType() && !initializerType.isa<FlipType>())`

That said, are nodes allowed on bundles with analog types?  If so, that will cause its own problem.  I wonder if this check is worth it for us: it would be straight-forward to just always call `convertToPassive` without the check, just like we do for primitives.

I'm ok with whatever you resolve here.

seldridge replied:
@clattner wrote:
> That said, are nodes allowed on bundles with analog types?

I checked... It looks like this is a weird edge case in the SFC:

1. Nodes cannot have analog type (at its outermost type)
2. Nodes must be passive (analog is passive as it's defined in CIRCT, too)
3. Attach can only involve ports, instance ports, or wires (this is the SFC's "kind" system rearing its head)

The combination of these three results in the inability to create a node that is analog, but the ability to create an aggregate node that contains analog. However, the kind checks disable any use of an analog in the node aggregate.

@jackkoenig: do you have any thoughts on why checks (1) and (3) exist?

I think a reasonable solution for us is to continue to do what we're doing: make nodes respect (2) and ignore (1) and (3). We'll accept more than the SFC, but I think that's okay. I think this is already what we're doing anyway. Alternatively, it's pretty easy to add the check of (1) if we want it. (3) is way trickier and should be dealt with during parsing if at all. (Leaning towards ignoring this.)

@clattner wrote:
> ... it would be straight-forward to just always call convertToPassive without the check, just like we do for primitives.
That would work, too. I'd prefer to accept more circuits than the SFC, but just minimally more. I'll come up with some solution. (Leaning towards leaving the check in).

seldridge replied:
And, good catch about how to simplify the logic based on type canonicalization. :+1: Using this now.

I've updated the commit to add a check of (1) during parsing. (1) is allowed once you're inside CIRCT, however. (3) is never checked.

jackkoenig replied:
The fact that you can have `Analog` as part of an aggregate node sounds like an oversight. The basic reasoning (which may not be relevant in CIRCT) is that there is directionality in node definitions. The flow is one way from right-hand side to the node at definition, and then after that you cannot connect to or otherwise "drive" a node. Since `attach` is a bidirectional operation, it cannot operate on nodes, thus it also doesn't make sense to have `Analog` in nodes. I guess you could still disallow (3) while allowing Analog in nodes though.

lattner replied:
> I think a reasonable solution for us is to continue to do what we're doing: make nodes respect (2) and ignore (1) and (3). We'll accept more than the SFC, but I think that's okay. I think this is already what we're doing anyway. 

I completely agree with you.

lattner reviewed (approved):
The patch looks great, with one more thing to think about.  I'm ok however you resolve it, you don't need another LGTM from me :-)
